src/Qir/Runtime/lib/QIR/QubitManager.hpp

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Qubit Manager maintains mapping between user qubit objects and
+// underlying qubit identifiers (Ids). When user program allocates
+// a qubit, Qubit Manager decides whether to allocate a fresh id or
+// reuse existing id that was previously freed. When user program
+// releases a qubit, Qubit Manager tracks it as a free qubit id.
+// Decision to reuse a qubit id is influenced by restricted reuse
+// areas. When a qubit id is freed in one section of a restricted
+// reuse area, it cannot be reused in other section of the same area.
+// Borrowing of qubits is not supported and is currently implemented
+// as plain allocation.
+
+#include <iostream>
+#include <vector>
+#include <limits>
+
+#include "CoreTypes.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+
+    // TODO: Remove this! Use standard ones.
+    static void FailNow(const char* message) { std::cout << message; throw message; }
+    static void FailIf(bool condition, const char* message) {
+        if (condition) {
+            FailNow(message);
+        }
+    }
```

Comment:
FYI:
* If a global function (as opposed to member function) is fully _defined_ in a header, then every .c/.cpp file that includes that header, will have a separate copy of such a function in the .obj file. Later on the linker is not guaranteed to merge the multiple copies into a single one (especially if the function is `static`).  
So, for global functions, you may want to leave in the header the _declaration_ only: `void FailNow(const char* message)` (optionally prepended with `extern`), and move the _definition_ to the .cpp file.
* I see that these particular functions are only used in one .cpp file. So it makes sense to completely move them from the header to that .cpp file and leave them `static`.
* Instead of `FailNow()` you may want to use [`quantum__rt__fail_cstr()`](https://github.com/microsoft/qsharp-runtime/blob/b10447eff672b7ba4f2e8564302888a8eb6bce8a/src/Qir/Runtime/lib/QIR/utils.cpp#L60).
* If you are interested about the future of `quantum__rt__fail_cstr()`, you can search for "quantum__rt__fail_cstr()" [here](https://ms-quantum.visualstudio.com/Quantum%20Program/_git/Quantum-OKRs.wiki?path=%2FMicrosoft-Quantum-Program%2FTechnical%2FEngineering-Specifications%2FCompiler-%26-Runtime%2FQIR-RT-API-Design-Guidelines.md&version=GBrokuzmin%2FqirRtApiGuidelines&_a=preview) (the doc is under development).

Reply:
This was obviously an integration point to be fixed with a TODO in front of it. I think the best way for now it to define these functions with quantum__rt__fail_cstr() as you suggested and move them to the .cpp file. That's what I did.